### ISSUE TITLE
ireg --repl -R <root-random-resource> assigns random storage resource to registered replica

### ISSUE THREAD

**Issue by tedgin** [irods/irods_client_icommands#230]
I'm using iRODS 4.2.8 on CentOS 7 servers. My iRODS zone has the following resource hierarchies.

```
CyVerseRes:random
├── ds01Broker:passthru
│   └── ds01:unixfilesystem
...
├── ds20Broker:passthru
│   └── ds20:unixfilesystem
├── dustyBroker:passthru
│   └── dusty:unixfilesystem
└── snoopyBroker:passthru
    └── snoopy:unixfilesystem

taccCorralRes:passthru
└── tacc:unixfilesystem
```
I have a data object named `/iplant/home/ipc_admin/hi.txt` with its only replica on `taccCorralRes;tacc`. I also have a file on `CyVerseRes;dustyBroker;dusty` that is a copy of this replica, and I want to register it as another replica of the associated data object. I use `ireg --repl` specifying the resource as `dusty`'s root resource `CyVerseRes`. Here's the command.

```sh
$ ireg --repl -R CyVerseRes /irods_vault/dusty_vault/home/ipc_admin/hi.txt /iplant/home/ipc_admin/hi.txt
```

The command succeeds, but surprisingly, instead of registering the file as a replica on `dusty` it chooses the peer storage resource `ds20`.

```sh
$ ils -L /iplant/home/ipc_admin/hi.txt
  ipc_admin         1 taccCorralRes;tacc            3 2017-04-22.15:09 & hi.txt
    764efa883dda1e11db47671c4a3bbd9e    text    /corral-tacc/projects/iplant/Vault/home/ipc_admin/hi.txt
  ipc_admin         2 CyVerseRes;ds20Broker;ds20            3 2021-09-23.12:59 & hi.txt
    764efa883dda1e11db47671c4a3bbd9e    text    /irods_vault/dusty_vault/home/ipc_admin/hi.txt
```

I verified that it didn't copy the file to the stated location `/irods_vault/dusty_vault/home/ipc_admin/hi.txt` on `ds20`'s server.

`ireg` should have either registered the file as a replica on `dusty` or generated on error.

**Comment by trel** [comment]
I think this is already handled in 4.2.9... we did a bunch of tightening up about the logic around replicas and their root/leafiness...   @alanking ?

Should be easy enough to reproduce...

**Comment by trel** [comment]
@tedgin what happens if you run the same `ireg` but name `dusty` directly?   Or the full hierarchy?

```
$ ireg --repl -R dusty /irods_vault/dusty_vault/home/ipc_admin/hi.txt /iplant/home/ipc_admin/hi.txt
OR
$ ireg --repl -R CyVerseRes;dustyBroker;dusty /irods_vault/dusty_vault/home/ipc_admin/hi.txt /iplant/home/ipc_admin/hi.txt
```

---

It looks like this particular case has a history...

There is a test that is now just a `pass`...

https://github.com/irods/irods/blob/6420dbdc0acd58e18fec98e1390678edf9b32df2/scripts/irods/test/test_ireg.py#L147-L149

It references https://github.com/irods/irods/issues/3517 which required targeting of a leaf resource in 4.1.11 but then is revisited in https://github.com/irods/irods/issues/3844 which allows root resources to be named (and then vote) for `ireg -R` in 4.2.3.

The voting mechanism worked, in this `random` case, but arguably should have a bit more intelligence in the case of the `--repl` flag that compares/checks the physical path on that resource?

Regardless, I think the workaround here (because `random` is non-deterministic, by design) is to just pass the leaf resource `dusty`, and I expect it will work as expected/desired.

We expect an administrator to use `ireg --repl` as surgically as possible as it deals with physical files already at rest in a specific place.

**Comment by alanking** [comment]
I have confirmed that the workaround works in 4.2.10 as well as 4.2.8 both in and outside of the vault:
```bash
$ ils -L foo
  rods              0 pt4;ufs4          283 2021-09-24.14:20 & foo
        generic    /var/lib/irods/ufs4vault/home/rods/foo
$ ls -l /var/lib/irods/ufs3vault/home/rods/foo
-rw------- 1 irods irods 283 Sep 24 14:20 /var/lib/irods/ufs3vault/home/rods/foo
$ ireg --repl -R ufs3 /var/lib/irods/ufs3vault/home/rods/foo /tempZone/home/rods/foo
$ ils -L foo
  rods              0 pt4;ufs4          283 2021-09-24.14:20 & foo
        generic    /var/lib/irods/ufs4vault/home/rods/foo
  rods              1 randy;pt3;ufs3          283 2021-09-24.14:22 & foo
        generic    /var/lib/irods/ufs3vault/home/rods/foo
```
```bash
$ ils -L foo
  rods              0 pt4;ufs4          283 2021-09-24.14:33 & foo
        generic    /var/lib/irods/ufs4vault/home/rods/foo
$ ls -l /var/lib/irods/notavault/foo
-rw------- 1 irods irods 283 Sep 24 14:35 /var/lib/irods/notavault/foo
$ ireg --repl -R ufs2 /var/lib/irods/notavault/foo /tempZone/home/rods/foo
$ ils -L foo
  rods              0 pt4;ufs4          283 2021-09-24.14:33 & foo
        generic    /var/lib/irods/ufs4vault/home/rods/foo
  rods              1 randy;pt2;ufs2          283 2021-09-24.14:36 & foo
        generic    /var/lib/irods/notavault/foo
```

**Comment by alanking** [comment]
Sorry, kinda slow today. The original issue still exists in 4.2.10:
```bash
$ ils -L foo
  rods              0 pt4;ufs4          283 2021-09-24.15:26 & foo
        generic    /var/lib/irods/ufs4vault/home/rods/foo
$ ls -l /var/lib/irods/ufs2vault/home/rods/foo
-rw------- 1 irods irods 283 Sep 24 15:26 /var/lib/irods/ufs2vault/home/rods/foo
$ ireg --repl -R randy /var/lib/irods/ufs2vault/home/rods/foo /tempZone/home/rods/foo
$ ils -L foo
  rods              0 pt4;ufs4          283 2021-09-24.15:26 & foo
        generic    /var/lib/irods/ufs4vault/home/rods/foo
  rods              1 randy;pt0;ufs0          283 2021-09-24.15:27 & foo
        generic    /var/lib/irods/ufs2vault/home/rods/foo
```
The `ireg` help text says:
> Register a file or a directory of files and subdirectory into iRODS.
> **The file or the directory of files must already exist on the server where
> the resource is located.**  The full path must be supplied for both the
> physicalFilePath and the irodsPath.

Right now, if the stat check fails, `--repl` code path inherits the system metadata from an existing sibling replica. This has been captured in irods/irods#4206.

irods/irods#3844 restored targeting coordinating resources for `ireg --repl` without explanation. So perhaps the question is: should this actually be *disallowed*? Can you explain the use case for targeting coordinating resources with `ireg --repl`?

**Comment by tedgin** [comment]
I don't have a use case for targeting a coordinating resource with `ireg --repl`.  It was a happy accident. I was just reporting the bad outcome as a bug. I'm happy with `ireg --repl` requiring the destination resource to be a storage resource as long as it fails with an error message and doesn't create an invalid data replica entry in the ICAT.

**Comment by alanking** [comment]
Okay, gotcha. Thanks, @tedgin!

@trel, perhaps we need to take a pass over `ireg` in much the same way that we put the scope on `irepl` and `iphymv` in the run up to 4.2.9 to make sure everything goes as expected.

**Comment by trel** [comment]
Yes, I'm thinking we revert allowing a root resource in this case.

Surgical tools should require precision of use.

@jasoncoposky do you remember any particulars of why https://github.com/irods/irods/issues/3844 was added back?   why wouldn't someone always specify a leaf?
